### PR TITLE
encoding of 0 should be the byte 0 and not the empty buffer.

### LINKED ISSUE #28: Bug. Decoding 0x80 / Encoding and decoding of 0
The rlp encoding of 0 is 0x80.

decoding 0x80 should return 0x0. The source returns an empty buffer which is incorrect.

`rlp.decode(rlp.encode(0))` returns `Buffer.from([])` instead of 0.

### PR DESCRIPTION
**Pull request by hamdiallam**
As stated in the spec,

`For a single byte whose value is in the [0x00, 0x7f] range, that byte is its own RLP encoding.`

Therefore, `rlp.encode(0)` should return `<Buffer 00>`. 

Fixes #28.

### CODE REVIEW

**Comment by hamdiallam** [comment]
The tests will fail due to the `ethereumjs-testing` has a bad test case RLP. A PR to fix that must be merged first.

File: index.js [hunk 1]
```diff
@@ -213,7 +213,7 @@ function toBuffer (v) {
       }
     } else if (typeof v === 'number') {
       if (!v) {
-        v = Buffer.from([])
+        v = Buffer.alloc(1)
```
Comment (axic) [inline index.js]: Either specify the value explicitly (such as `from([0])`) or use `allocSafe` because that initialises with zero.

Comment (hamdiallam) [reply index.js]: got it.

**Comment by axic** [comment]
@holgerd77 after https://github.com/ethereum/tests/pull/478 is merged we need to update our `ethereumjs-testing` repo in order to merge this PR

**Comment by holgerd77** [comment]
@axic I was left a bit confused on this. I read a bit through the discussion on the tests repo, am I correct that this was falsely identified as a bug or at least not be the current/common state of the implementations? So should this be closed?

**Comment by hamdiallam** [comment]
As far as I know, this is still bug that hasn't been fixed. Tests are failing due to the PR is the testing repo not being merged

**Comment by s1na** [comment]
The upstream test [PR](https://github.com/ethereum/tests/pull/478) has been rolled back in https://github.com/ethereum/tests/pull/485.

> RLP encoding of 0 is encoding of empty byte array, so 0x80 is correct.
> 0x00 is encoding of byte array of length 1 containing one byte 0, not encoding of integer 0.

Can we close this PR?

**Comment by holgerd77** [comment]
Do we have/can we add some place where we have an explicit note on this, being it a comment on a test or some small documentation note before closing?

File: test/index.js [hunk 2]
```diff
@@ -75,7 +75,7 @@ describe('RLP encoding (integer):', function () {
   })
 
   it('it should handle zero', function () {
-    assert.equal(RLP.encode(0).toString('hex'), '80')
+    assert.equal(RLP.encode(0).toString('hex'), '00')
```
Comment (s1na) [inline test/index.js]: @hamdiallam Can you please add a small comment, maybe mentioning the link to this PR (or #28) and revert the change in `index.js` so we can merge this instead of closing it?

Of course if you're still in doubt about whether `0x80` is the correct encoding we can keep the discussion open.

Comment (hamdiallam) [reply test/index.js]: I'm confused as to why the encoding of 0 should be '80', given the input is a single byte.

Inconsistent with this statement in the encoding specification:
`For a single byte whose value is in the [0x00, 0x7f] range, that byte is its own RLP encoding.`

Comment (hamdiallam) [reply test/index.js]: I see the comment made being that `0` is an empty byte array. Is this expected behavior? I'd assume `''` or `[]` is an empty byte array but not `0`

**Comment by hamdiallam** [comment]
I see the comments made. Continuing conversation in [ethereum/tests#485](https://github.com/ethereum/tests/pull/485).

**Comment by simon-jentzsch** [comment]
As far as I see it:

The RLP-Spec itself does not define how to convert a number in its bytes-representation. That's why we are having this discussion.
But Ethereum does. And there the former implementation, that uses a empty Array is the correct way.

I also use rlp.encode(txIndex) when creating the key for the merkleTrie. And currently the works correct, since txIndex=0 is converted to an empty Buffer, but this PR would break my (and other peoples) code!

**Comment by axic** [comment]
Can the YP be updated to codify this behaviour?

**Comment by ryanio** [comment]
Since https://eth.wiki/en/fundamentals/rlp now has these examples I will consider this conversation resolved:

>The integer 0 = `[ 0x80 ]`
> The encoded integer 0 (’\x00’) = `[ 0x00 ]`

Thanks to everyone for participating

**Comment by holgerd77** [comment]
Small note: we should update the URL on the README to directly point to the eth.wiki URL.

**Comment by ryanio** [comment]
@holgerd77 yes for sure, I already have it updated in PR 104, I'll make sure it's included as part of v3 release.
